### ISSUE TITLE
Build failure with clang: object of type ... cannot be assigned because its copy assignment operator is implicitly deleted

### ISSUE THREAD

**Issue by ryandesign** [mbunkus/mkvtoolnix#1175]
Hello, I'm a developer with the MacPorts project. mkvtoolnix fails to build with the clang compiler on OS X. This is the error with mkvtoolnix 7.8.0 (though we've been [seeing this error for years in MacPorts](https://trac.macports.org/ticket/40231):

```
In file included from src/common/command_line.cpp:14:
In file included from src/common/common_pch.h:4:
In file included from src/common/common.h:29:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/vector:1950:21: fatal error: object of type 'boost::iterators::filter_iterator<(lambda at src/common/command_line.cpp:301:74), boost::iterators::transform_iterator<boost::range_detail::select_first<std::__1::map<std::__1::basic_string<char>, std::__1::basic_string<char>, std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, std::__1::basic_string<char> > > > >, std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char>, std::__1::basic_string<char> >, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char>, std::__1::basic_string<char> >, void *> *, long> >, boost::iterators::use_default, boost::iterators::use_default> >' cannot be assigned because its copy assignment operator is implicitly deleted
                __m = __first;
                    ^
/opt/local/include/boost/range/algorithm_ext/push_back.hpp:33:8: note: in instantiation of function template specialization 'std::__1::vector<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >::insert<boost::iterators::filter_iterator<(lambda at src/common/command_line.cpp:301:74), boost::iterators::transform_iterator<boost::range_detail::select_first<std::__1::map<std::__1::basic_string<char>, std::__1::basic_string<char>, std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, std::__1::basic_string<char> > > > >, std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char>, std::__1::basic_string<char> >, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char>, std::__1::basic_string<char> >, void *> *, long> >, boost::iterators::use_default, boost::iterators::use_default> > >' requested here
    on.insert( on.end(), boost::begin(from), boost::end(from) );
       ^
src/common/command_line.cpp:301:13: note: in instantiation of function template specialization 'boost::range::push_back<std::__1::vector<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >, boost::range_detail::filtered_range<(lambda at src/common/command_line.cpp:301:74), const boost::range_detail::select_first_range<std::__1::map<std::__1::basic_string<char>, std::__1::basic_string<char>, std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, std::__1::basic_string<char> > > > > > >' requested here
      brng::push_back(keys, rel.urls | badap::map_keys | badap::filtered([](std::string const &key) { return key != "general"; }));
            ^
/opt/local/include/boost/iterator/filter_iterator.hpp:107:17: note: copy assignment operator of 'filter_iterator<(lambda at src/common/command_line.cpp:301:74), boost::iterators::transform_iterator<boost::range_detail::select_first<std::__1::map<std::__1::basic_string<char>, std::__1::basic_string<char>, std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, std::__1::basic_string<char> > > > >, std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char>, std::__1::basic_string<char> >, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char>, std::__1::basic_string<char> >, void *> *, long> >, boost::iterators::use_default, boost::iterators::use_default> >' is implicitly deleted because field 'm_predicate' has a deleted copy assignment operator
      Predicate m_predicate;
                ^
src/common/command_line.cpp:301:74: note: lambda expression begins here
      brng::push_back(keys, rel.urls | badap::map_keys | badap::filtered([](std::string const &key) { return key != "general"; }));
                                                                         ^
```

This is the version of the clang compiler I am using:

```
$ clang -v
Apple LLVM version 6.1.0 (clang-602.0.49) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.3.0
Thread model: posix
```

This is the version of Xcode:

```
$ xcodebuild -version
Xcode 6.3.1
Build version 6D1002
```

This is the version of OS X:

```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.10.3
BuildVersion:	14D136
```

We are working around this problem in MacPorts by [commenting out](https://trac.macports.org/browser/trunk/dports/multimedia/mkvtoolnix/files/patch-clang_errors.diff?rev=120972) the portion of src/common/command_line.cpp that is bracketed by `#if defined(HAVE_CURL_EASY_H)` ... `#endif` (the part that handles the `--check-for-updates` command line argument) but I'm sure there's a better fix possible.

**Comment by mbunkus** [comment]
I'm curious what makes Homebrew so special. I haven't had any such error on Linux with clang 3.6.0 + gcc's libstdc++, nor on Mac OS outside of Homebrew (Xcode 6.2 on OS X 10.9, clang 3.5.0 with clang's libc++). Sure, I can apply your workaround, but I'd like to understand the underlying issue first.

If you're commenting out that block then you can also disable the use of curl completely with the corresponding option to configure, `--without-curl`.

**Comment by mbunkus** [comment]
Some more information. I've built MKVToolNix on Mac with the aforementioned XCode myself just last weekend. I know from my Mac DMG maintainer that he's using XCode 6.3 on Mac 10.10.3 without additional patches himself.

Both him and me are compiling the required libraries by explicitly doing `export CXXFLAGS="-std=c++11 -stdlib=libc++"` for the libraries and `export CXXFLAGS="-stdlib=libc++"` for MKVToolNix.

It looks to me like your Boost might have been compiled with different settings (e.g. without `-std=c++11`) or with a different compiler (e.g. gcc instead of clang).

**Comment by mbunkus** [comment]
Also note that I don't support Mac OS myself. I don't have a Mac, only have partial access to one, have never developed on that platform and have no interest at all in spending time on it, especially not on such special problems if compilation generally succeeds on that platform.

**Comment by mbunkus** [comment]
Ooops, I meant »what makes MacPorts so special«, not Homebrew. Sorry for confusing the two projects.

**Comment by mojca** [comment]
I wouldn't be surprised if building boost with or without `export CXXFLAGS="-std=c++11"` accounts for the difference.  We can test that and also check how HomeBrew developers build dependencies. It always helps to know that one solution works out-of-the-box and then compare what's different in the other setup.

(Proper `stdlib` setting is already default and in place on 10.9 and later. On 10.8 and below we are aware that our setup is broken, but that's a different issue altogether; something that we still need to fix.)

No problem about mixing the projects. They all serve the same purpose. In fact it is often very helpful when one community solves a particular problem and then the other community can "cheat" and replicate the solution.

**Comment by ryandesign** [comment]
When we encounter problems with the software packages we've added to MacPorts, if we believe the problems are not MacPorts-specific, we usually prefer to report those problems to the developers of the software, so that they can be fixed at the original source, so that the fix can benefit all users, not just MacPorts users. Usually, the MacPorts contributor who has volunteered to be the maintainer of the package would do that, but mkvtoolnix has not had a MacPorts maintainer in awhile, so it has had a few patches added to it over the years that were not sent back to you, a situation which I'm now trying to remedy.

Thanks for letting me know that you've been able to build mkvtoolnix on OS X without patches; I assumed the problems we encountered in MacPorts were also being encountered by other OS X users. Building within MacPorts is usually not very different from building outside of MacPorts, though MacPorts does set a number of environment variables that help things build consistently. I will try building mkvtoolnix outside of MacPorts to see if I can reproduce your success.

We're very aware in MacPorts of the different compilers and the different C++ libraries and the importance of using the same C++ library for everything, so I don't believe that is related to the problem I'm experiencing; I've just double-checked and confirmed that on my system mkvtoolnix and all its dependencies are linked only with libc++, not libstdc++.

`-stdlib=libc++` is the default on OS X 10.9 and later so that should have no effect. We are not specifically adding `-std=c++11` to the build of mkvtoolnix's dependencies; that's not something we typically add in MacPorts. If a particular project required it, usually we would expect the project's build system to add it.

Also, just to clarify, I'm not suggesting you should disable this block of code; I'm just explaining that's what one of our developers did two years ago to avoid the problem; I thought this might help you localize which part of the code was related to the problem.

**Comment by mbunkus** [comment]
Adding `-std=c++11` to the `CXXFLAGS` for MKVToolNix would actually be wrong. configure detects which features the compiler supports and also probes for C++14 support. XCode 6.3 with its clang 3.6.0 does already support `-std=c++14`, and configure will use that automatically.

The problem with forcing `CXXFLAGS=-std=c++11` outside of configure is that configure thinks C++14 support is available and therefore certain workaround code is not built. But your manual `CXXFLAGS` override the detected ones during compilation, and then compilation will abort.

Just to clarify: don't add `-std=c++11` for MKVToolNix. I was only providing as much information as possible on how I compiled the other libraries.

I do want bug reports upstreamed (= reported here), definitely. In this case I would like to understand what the problem is, because unlike #1176 which definitely is a programming error this looks more like an issue in either a third-party library or in the compiler settings. Of course I can write that code differently (e.g. without the use of Boost's range adapters), but then I will most likely repeat such problems in the future. For me it's also about preventing such issues again.

**Comment by mojca** [comment]
Again just to clarify: we currently don't add any `CXXFLAGS=-std=c++11` flags to any package (neither MKVToolNix nor ICU/Boost). You said you set them when building ICU and Boost. I was just suggesting that maybe we should also try setting those flags on ICU/Boost as an experiment as I wouldn't be surprised if the results (success vs. failure) were different. But thanks for clarifying that we should not attempt to do the same with MKVToolNix itself. That might save us one build cycle.

**Comment by mbunkus** [comment]
I apologize: I have to take back parts of what I said. My builds on Mac were done without curl, therefore the problematic section wasn't even compiled. I've just tried this with curl enabled and I run into the same problem here with XCode 6.2.

I see in the [Homebrew formula for MKVToolNix](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/mkvtoolnix.rb) that they compile with curl explicitly disabled, too.

I'll implement a workaround for this piece of code even though I don't understand yet what the underlying problem actually is (or why it only occurs on Mac with clang but not on e.g. Linux with clang).

**Comment by mbunkus** [comment]
I found [this interesting post on StackOverflow](http://stackoverflow.com/questions/25906893/filtered-ranges-lambdas-and-is-sorted) about this very problem. Funny – my browser history shows that I've actually visited it before, so I guess I've had a similar problem in the past.

Both of the fixes mentioned in the accepted answer work for our problem. I'll continue compilation and will push a fix if everything succeeds.
